File: smartlist/src/main/java/com/smartlist/module/user/UserController.java
package com.smartlist.module.user;

import com.smartlist.module.user.dto.request.*;
import com.smartlist.module.user.dto.response.UserRespDTO;
import com.smartlist.utils.AppResponseCodes;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.media.Content;
import io.swagger.v3.oas.annotations.media.Schema;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.responses.ApiResponses;
import io.swagger.v3.oas.annotations.tags.Tag;
import jakarta.validation.Valid;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Positive;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.security.access.annotation.Secured;
import org.springframework.validation.annotation.Validated;
import org.springframework.web.bind.annotation.*;

@RestController
@Tag(name = "User", description = "User API")
@RequestMapping(value ="/user/", produces = { "application/json" })
@Validated
@Secured({"SUPERADMIN", "ADMIN", "MEMBER"})
public class UserController {

    private final UserService userService;

    @Autowired
    public UserController(UserService userService){
        this.userService = userService;
    }

    @Operation(summary = "Create a new user")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", schema = @Schema(implementation = Integer.class)) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = UserResponseCodes.USER_EMAIL_ID_EXISTS_DESCRIPTION, content = @Content)
    })
    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping("create")
    public Integer create(@RequestBody @Valid CreateUserReqDTO createUserReqDto) {
        return this.userService.create(createUserReqDto);
    }

    @Operation(summary = "Create a new user from email invite")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", schema = @Schema(implementation = Integer.class)) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = UserResponseCodes.TOKEN_EXPIRED_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = UserResponseCodes.INVALID_TOKEN_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "404", description = UserResponseCodes.USER_EMAIL_ID_EXISTS_DESCRIPTION, content = @Content)
    })
    @ResponseStatus(HttpStatus.CREATED)
    @PostMapping("create-from-invite")
    public Integer createFromInvite(@RequestBody @Valid CreateUserFromInviteReqDTO createUserFromInviteReqDto){
        return this.userService.createFromInvite(createUserFromInviteReqDto);
    }

    @Operation(summary = "Verify new user email")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", schema = @Schema(implementation = Integer.class)) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = UserResponseCodes.TOKEN_EXPIRED_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = UserResponseCodes.INVALID_TOKEN_DESCRIPTION, content = @Content)
    })
    @PutMapping("verify-email")
    public void verifyEmail(@RequestParam(value = "token") @NotBlank String token) {
       this.userService.verifyUserEmail(token);
    }

    @Operation(summary = "Regenerate email verification")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content =  @Content),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = UserResponseCodes.INVALID_USER_EMAIL_ID_DESCRIPTION, content = @Content)
    })
    @PutMapping("regenerate-verify-email")
    public void regenerateEmailVerification(@RequestBody @Valid RegenerateVerificationEmailReqDTO regenerateVerificationEmailReqDto) {
        this.userService.regenerateEmailVerification(regenerateVerificationEmailReqDto.getEmail());
    }

    @Operation(summary = "Forgot Password")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content =  @Content),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = UserResponseCodes.INVALID_USER_EMAIL_ID_DESCRIPTION, content = @Content)
    })
    @PostMapping("forgot-password")
    public void forgotPassword(@RequestBody @Valid ForgotPasswordReqDTO forgotPasswordReqDto) {
        this.userService.verifyAndSendForgotPasswordEmail(forgotPasswordReqDto.getEmail());
    }

    @Operation(summary = "Reset Forgot Password")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content =  @Content),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = UserResponseCodes.TOKEN_EXPIRED_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = UserResponseCodes.INVALID_TOKEN_DESCRIPTION, content = @Content)
    })
    @PostMapping("forgot-password-reset")
    public void forgotPasswordReset(@RequestBody @Valid ForgotPasswordResetReqDTO forgotPasswordResetReqDto) {
        this.userService.verifyAndResetPassword(forgotPasswordResetReqDto);
    }

    @Operation(summary = "Change password")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content =  @Content),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = UserResponseCodes.INVALID_PASSWORD_DESCRIPTION, content = @Content)
    })
    @PostMapping("{id}/change-password")
    public void changePassword(@PathVariable("id") @Positive Integer id, @RequestBody() @Valid ChangePasswordReqDTO changePasswordReqDto) {
        this.userService.changePassword(id, changePasswordReqDto);
    }

    @Operation(summary = "Get user")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content = { @Content(mediaType = "application/json", schema = @Schema(implementation = UserRespDTO.class)) }),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = UserResponseCodes.INVALID_USER_ID_DESCRIPTION, content = @Content)
    })
    @GetMapping("{id}")
    public UserRespDTO findOne(@PathVariable("id") @Positive Integer id) {
        return this.userService.findUserById(id);
    }

    @Operation(summary = "Update user")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content =  @Content),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = UserResponseCodes.INVALID_PASSWORD_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = UserResponseCodes.INVALID_PROFILE_PHOTO_FILE_DESCRIPTION, content = @Content)
    })
    @PutMapping("{id}")
    public void update(@PathVariable("id") @Positive Integer id, @RequestBody() @Valid UpdateUserReqDTO updateUserReqDto) {
        this.userService.updateUser(id, updateUserReqDto);
    }

    @Operation(summary = "Toggle user's active status")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content =  @Content),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = UserResponseCodes.INVALID_USER_ID_DESCRIPTION, content = @Content)
    })
    @PatchMapping("{id}/active/{status}")
    public void setUserActive(@PathVariable("id") @Positive Integer userId, @PathVariable("status") @NotNull boolean status) {
        this.userService.toggleUserActive(userId, status);
    }

    @Operation(summary = "Update user's organization by id")
    @ApiResponses(value = {
            @ApiResponse(responseCode = "200", content =  @Content),
            @ApiResponse(responseCode = "400", description = AppResponseCodes.BAD_REQUEST_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = UserResponseCodes.INVALID_USER_ID_DESCRIPTION, content = @Content),
            @ApiResponse(responseCode = "400", description = UserResponseCodes.INVALID_ORGANIZATION_ID_DESCRIPTION, content = @Content)
    })
    @PatchMapping("{id}/organization/{organizationId}")
    public void updateUserOrganization(@PathVariable("id") @Positive Integer id, @PathVariable("organizationId") @Positive Integer organizationId) {
        this.userService.updateUserOrganization(id, organizationId);
    }
}

File: javascript/smartlist/smartlist-api-mysql-project/src/modules/user/user.controller.ts
import {
  Body,
  Controller,
  Get,
  HttpCode,
  Param,
  ParseBoolPipe,
  ParseIntPipe,
  Patch,
  Post,
  Put,
  Query,
} from '@nestjs/common';
import {
  ApiOperation,
  ApiProduces,
  ApiResponse,
  ApiTags,
} from '@nestjs/swagger';

import { CreateUserReqDto } from './dto/request/create-user.req.dto';
import { ForgotPasswordResetReqDto } from './dto/request/forgot-password-reset.req.dto';
import { ForgotPasswordReqDto } from './dto/request/forgot-password.req.dto';
import { ChangePasswordReqDto } from './dto/request/change-password.req.dto';
import { UpdateUserReqDto } from './dto/request/update-user.req.dto';
import { UserService } from './user.service';
import { UserRespDto } from './dto/response/user.resp.dto';
import { CreateUserFromInviteReqDto } from './dto/request/create-user-from-invite.req.dto';
import { UserResponseCodes } from './user.response.codes';
import { RegenerateVerificationEmailReqDto } from './dto/request/regenerate-verification-email.req.dto';
import { Public } from '../auth/jwt-auth.guard';
import { Role } from 'src/enums/role';
import { Roles } from 'src/guards/roles.decorator';

@Controller('/user')
@Roles(Role.SUPERADMIN, Role.ADMIN, Role.MEMBER)
@ApiTags('User')
export class UserController {
  constructor(private readonly userService: UserService) {}

  @ApiOperation({ summary: 'Create a new user' })
  @ApiProduces('application/text')
  @ApiResponse({ ...UserResponseCodes.CREATED, type: Number })
  @ApiResponse(UserResponseCodes.BAD_REQUEST)
  @ApiResponse(UserResponseCodes.USER_EMAIL_ID_EXISTS)
  @Public()
  @HttpCode(201)
  @Post('create')
  create(@Body() createUserReqDto: CreateUserReqDto): Promise<number> {
    return this.userService.create(createUserReqDto);
  }

  @ApiOperation({ summary: 'Create a new user from email invite' })
  @ApiProduces('application/text')
  @ApiResponse({ ...UserResponseCodes.CREATED, type: Number })
  @ApiResponse(UserResponseCodes.BAD_REQUEST)
  @ApiResponse(UserResponseCodes.TOKEN_EXPIRED)
  @ApiResponse(UserResponseCodes.USER_EMAIL_ID_EXISTS)
  @ApiResponse(UserResponseCodes.INVALID_TOKEN)
  @Public()
  @HttpCode(201)
  @Post('create-from-invite')
  createFromInvite(
    @Body() createUserFromInviteReqDto: CreateUserFromInviteReqDto,
  ): Promise<number> {
    return this.userService.createFromInvite(createUserFromInviteReqDto);
  }

  @ApiOperation({ summary: 'Verify new user email' })
  @ApiResponse(UserResponseCodes.SUCCESS)
  @ApiResponse(UserResponseCodes.BAD_REQUEST)
  @ApiResponse(UserResponseCodes.TOKEN_EXPIRED)
  @ApiResponse(UserResponseCodes.INVALID_TOKEN)
  @Public()
  @HttpCode(200)
  @Put('verify-email')
  verifyEmail(@Query('token') token: string) {
    return this.userService.verifyUserEmail(token);
  }

  @ApiOperation({ summary: 'Regenerate email verification' })
  @ApiResponse(UserResponseCodes.SUCCESS)
  @ApiResponse(UserResponseCodes.BAD_REQUEST)
  @ApiResponse(UserResponseCodes.INVALID_USER_EMAIL_ID)
  @Public()
  @HttpCode(200)
  @Put('regenerate-verify-email')
  regenerateEmailVerification(
    @Body()
    regenerateVerificationEmailReqDto: RegenerateVerificationEmailReqDto,
  ) {
    return this.userService.regenerateEmailVerification(
      regenerateVerificationEmailReqDto.email,
    );
  }

  @ApiOperation({ summary: 'Forgot Password' })
  @ApiResponse(UserResponseCodes.SUCCESS)
  @ApiResponse(UserResponseCodes.BAD_REQUEST)
  @ApiResponse(UserResponseCodes.INVALID_USER_EMAIL_ID)
  @Public()
  @HttpCode(200)
  @Post('forgot-password')
  forgotPassword(@Body() forgotPasswordReqDto: ForgotPasswordReqDto) {
    return this.userService.verifyAndSendForgotPasswordEmail(
      forgotPasswordReqDto.email,
    );
  }

  @ApiOperation({ summary: 'Reset Forgot Password' })
  @ApiResponse(UserResponseCodes.SUCCESS)
  @ApiResponse(UserResponseCodes.BAD_REQUEST)
  @ApiResponse(UserResponseCodes.TOKEN_EXPIRED)
  @ApiResponse(UserResponseCodes.INVALID_TOKEN)
  @Public()
  @HttpCode(200)
  @Post('forgot-password-reset')
  forgotPasswordReset(
    @Body() forgotPasswordResetReqDto: ForgotPasswordResetReqDto,
  ) {
    return this.userService.verifyAndResetPassword(forgotPasswordResetReqDto);
  }

  @ApiOperation({ summary: 'Change password' })
  @ApiResponse(UserResponseCodes.SUCCESS)
  @ApiResponse(UserResponseCodes.BAD_REQUEST)
  @ApiResponse(UserResponseCodes.INVALID_USER_ID_OR_PASSWORD)
  @HttpCode(200)
  @Post(':id/change-password')
  changePassword(
    @Param('id', ParseIntPipe) id: number,
    @Body() changePasswordReqDto: ChangePasswordReqDto,
  ) {
    return this.userService.changePassword(id, changePasswordReqDto);
  }

  @ApiOperation({ summary: 'Get user' })
  @ApiResponse({ ...UserResponseCodes.SUCCESS, type: UserRespDto })
  @ApiResponse(UserResponseCodes.BAD_REQUEST)
  @ApiResponse(UserResponseCodes.INVALID_USER_ID)
  @HttpCode(200)
  @Get(':id')
  findOne(@Param('id', ParseIntPipe) id: number) {
    return this.userService.findUserById(id);
  }

  @ApiOperation({ summary: 'Update user' })
  @ApiResponse(UserResponseCodes.SUCCESS)
  @ApiResponse(UserResponseCodes.BAD_REQUEST)
  @ApiResponse(UserResponseCodes.INVALID_USER_ID)
  @ApiResponse(UserResponseCodes.INVALID__PROFILE_PHOTO_FILE_ID)
  @HttpCode(200)
  @Put(':id')
  update(
    @Param('id', ParseIntPipe) id: number,
    @Body() updateUserReqDto: UpdateUserReqDto,
  ) {
    return this.userService.updateUser(id, updateUserReqDto);
  }

  @ApiOperation({ summary: "Toggle user's active status" })
  @ApiResponse(UserResponseCodes.SUCCESS)
  @ApiResponse(UserResponseCodes.BAD_REQUEST)
  @ApiResponse(UserResponseCodes.INVALID_USER_ID)
  @HttpCode(200)
  @Patch(':id/active/:status')
  setUserActive(
    @Param('id', ParseIntPipe) userId: number,
    @Param('status', ParseBoolPipe) status,
  ) {
    return this.userService.toggleUserActive(userId, status);
  }

  @ApiOperation({ summary: "Update user's organization by id" })
  @ApiResponse(UserResponseCodes.SUCCESS)
  @ApiResponse(UserResponseCodes.BAD_REQUEST)
  @ApiResponse(UserResponseCodes.INVALID_USER_ID)
  @ApiResponse(UserResponseCodes.INVALID_ORGANIZATION_ID)
  @HttpCode(200)
  @Patch(':id/organization/:organizationId')
  updateUserOrganization(
    @Param('id', ParseIntPipe) id: number,
    @Param('organizationId', ParseIntPipe) organizationId: number,
  ) {
    return this.userService.updateUserOrganization(id, organizationId);
  }
}
